autorag/retriever/google_and_vector_retriever.py

```diff
@@ -0,0 +1,95 @@
+import os
+from llama_index.schema import TextNode, NodeWithScore, QueryBundle
+from llama_index.indices.vector_store.retrievers.retriever import VectorIndexRetriever
+from llama_index.core.base_retriever import BaseRetriever
+import requests
+from bs4 import BeautifulSoup
+from googleapiclient.discovery import build
+from typing import Any, Dict, List, Optional
+import re
+
+
+class GoogleRetriever(BaseRetriever):
+    """Custom retriever that performs both semantic search and hybrid search."""
+
+    def __init__(
+        self,
+        api_key: str = None,
+        cse_id: str = None,
+        topk: int = 10,
+    ) -> None:
+        """Init params."""
+        api_key = api_key or os.environ["GOOGLE_SEARCH_API_KEY"]
+        cse_id = cse_id or os.environ["GOOGLE_SEARCH_CSE_ID"]
+        self.service = build("customsearch", "v1", developerKey=api_key).cse()
+        self.cse_id = cse_id
+        self.topk = topk
+        super().__init__()
+
+    @staticmethod
+    def fetch_page_text(url):
+        try:
+            response = requests.get(url)
+            response.raise_for_status()  # Raises an HTTPError for bad responses
+            soup = BeautifulSoup(response.text, "html.parser")
+            text = soup.get_text()
+            text = re.sub(r"\s+", " ", text)
+            text = re.sub(r"\n+", "n", text)
```

Comment:
Replacing `r"\n+"` to` "n"` is is somewhat unusual in this context. should we change "n" to " "?

Reply:
good catch!

Reply:
will change to \n